### PR TITLE
*: propagate workdir to children containers

### PR DESCRIPTION
**Pull request by erizocosmico**
Since daemon itself was not going to do anything with the workdir decided not to mount it and just pass it as an argument. Children containers are started with the workdir mounted if needed.

Going forward containers may need some configuration and `InfoOrStart` provided too little possibility for configuration from the outside, so I added the possibility to pass it some config options.

### CODE REVIEW

File: cmd/srcd-server/engine/engine.go [hunk 1]
```diff
@@ -152,7 +154,11 @@ func (s *Server) parse(ctx context.Context, req *api.ParseRequest, log logf) (*a
 	return resp, nil
 }
 
-func createBbblfshd() error {
+func (s *Server) withWorkdirMounted(at string) docker.ConfigOption {
```
Comment (campoy) [inline cmd/srcd-server/engine/engine.go]: wondering whether this function is actually needed
it's only called once and the body is simple enough I would just manually inline it

File: cmd/srcd-server/engine/sql.go [hunk 2]
```diff
@@ -15,9 +15,14 @@ import (
 )
 
 const gitbaseName = "srcd-cli-gitbase"
+const gitbaseMountPath = "/opt/repos"
```
Comment (campoy) [inline cmd/srcd-server/engine/sql.go]: const (
   gitbaseName = ...
   gitbaseMountPath = ...
)

File: cmd/srcd/cmd/sql.go [hunk 3]
```diff
@@ -40,7 +40,12 @@ var sqlCmd = &cobra.Command{
 		}
 		query := args[0]
 
-		c, err := daemon.Client()
+		workdir, err := os.Getwd()
+		if err != nil {
+			logrus.Fatalf("could not get working directory: %s", err)
+		}
+
+		c, err := daemon.Client(workdir)
```
Comment (campoy) [inline cmd/srcd/cmd/sql.go]: I feel like you could create a new function daemon.ClientOnPwd and move the code calling os.Getwd into it

File: cmd/srcd/daemon/daemon.go [hunk 4]
```diff
@@ -42,8 +43,13 @@ func Client() (api.EngineClient, error) {
 	return api.NewEngineClient(conn), nil
 }
 
-func Start() error {
+func Start(workdir string) error {
+	return start(docker.WithEnv(workdirKey, workdir))
+}
+
+func start(opts ...docker.ConfigOption) error {
```
Comment (campoy) [inline cmd/srcd/daemon/daemon.go]: I'd make start(opts) return a `func() error` so you don't need to pass the options around so much.

Closures ftw

**Comment by erizocosmico** [comment]
Fixed

File: cmd/srcd/cmd/sql.go (hunk 3 above)
Comment (erizocosmico) [reply cmd/srcd/cmd/sql.go]: Actually, Client never needed workdir cause the only place you need a workdir different than pwd is Start, so I'm gonna remove all that

**Comment by erizocosmico** [comment]
Fixed the issue. Right now it's a little hacky, because it kills all components (which is all docker containers here). I will rebase #9 with this and make it only remove the components that are workdir dependant (no need to kill bblfsh, for example, which doesn't require workdir).

**Pull request merged**